File: Problems & Solutions/2246. Longest Path With Different Adjacent Characters/13 January 2022/Approach 2 - Breadth First Search/Longest Path With Different Adjacent Characters.java
class Solution {
    public int longestPath(int[] parent, String s) {
        int n = parent.length;
        int[] childrenCount = new int[n];
        for (int node = 1; node < n; node++) {
            childrenCount[parent[node]]++;
        }

        Queue<Integer> q = new LinkedList<>();
        int longestPath = 1;
        int[][] longestChains = new int[n][2];

        for (int node = 1; node < n; node++) {
            if (childrenCount[node] == 0) {
                longestChains[node][0] = 1;
                q.offer(node);
            }
        }

        while (!q.isEmpty()) {
            int currentNode = q.poll();
            int par = parent[currentNode];
            int longestChainStartingFromCurrNode = longestChains[currentNode][0];
            if (s.charAt(currentNode) != s.charAt(par)) {
                if (longestChainStartingFromCurrNode > longestChains[par][0]) {
                    longestChains[par][1] = longestChains[par][0];
                    longestChains[par][0] = longestChainStartingFromCurrNode;
                } else if (longestChainStartingFromCurrNode > longestChains[par][1]) {
                    longestChains[par][1] = longestChainStartingFromCurrNode;
                }
            }

            longestPath = Math.max(longestPath, longestChains[par][0] + longestChains[par][1] + 1);
            childrenCount[par]--;

            if (childrenCount[par] == 0 && par != 0) {
                longestChains[par][0]++;
                q.offer(par);
            }
        }

        return longestPath;
    }
}

File: Problems & Solutions/2246. Longest Path With Different Adjacent Characters/13 January 2022/Approach 1 - Depth First Search/Longest Path With Different Adjacent Characters.cpp
class Solution {
public:
    int dfs(int currentNode, vector<vector<int>>& children, string& s, int& longestPath) {
        int longestChain = 0, secondLongestChain = 0;
        for (int child : children[currentNode]) {
            int longestChainStartingFromChild = dfs(child, children, s, longestPath);
            if (s[currentNode] == s[child]) {
                continue;
            }
            if (longestChainStartingFromChild > longestChain) {
                secondLongestChain = longestChain;
                longestChain = longestChainStartingFromChild;
            } else if (longestChainStartingFromChild > secondLongestChain) {
                secondLongestChain = longestChainStartingFromChild;
            }
        }

        longestPath = max(longestPath, longestChain + secondLongestChain + 1);
        return longestChain + 1;
    }

    int longestPath(vector<int>& parent, string s) {
        int n = parent.size();
        vector<vector<int>> children(n);
        for (int i = 1; i < n; i++) {
            children[parent[i]].push_back(i);
        }

        int longestPath = 1;
        dfs(0, children, s, longestPath);

        return longestPath;
    }
};